### DISCUSSION TITLE
Why does CMake wrongly believe the rimage target is changing?

### DISCUSSION THREAD

**Discussion by marc-hb** [zephyrproject-rtos/zephyr#40781] [Q&A] [answered]
_Originally posted by @tejlmand in https://github.com/zephyrproject-rtos/zephyr/pull/40431#issuecomment-978007591_

> > Why does CMake wrongly believe the rimage target is changing?
>
>  You shouldn't be passing -- -DFOO=bar on subsequent builds:

That might be true when using `west`, because `west` would re-invoke `cmake`, but when re-running pure cmake directly give the same warning:
```
$ cmake .
...
CMake Warning at /projects/github/ncs/zephyr/cmake/extensions.cmake:2393 (message):
  The build directory must be cleaned pristinely when changing rimage target,

  Current value="", Ignored value="cnl"
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/extensions.cmake:829 (zephyr_check_cache)
  /projects/github/ncs/zephyr/boards/xtensa/intel_adsp_cavs18/board.cmake:6 (board_set_rimage_target)
  /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake:627 (include)
  /projects/github/ncs/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:24 (include)
  /projects/github/ncs/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:40 (include_boilerplate)
  CMakeLists.txt:5 (find_package)
```

which is very bad, cause a regular rebuild using `ninja` will trigger the same, if a cmake re-run is needed.

Just  try:
```
$ touch CMakeLists.txt
$ ninja -Cbuild
ninja: Entering directory `build'
[0/1] Re-running CMake...
Including boilerplate (Zephyr base (cached)): /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake
...
CMake Warning at /projects/github/ncs/zephyr/cmake/extensions.cmake:2393 (message):
  The build directory must be cleaned pristinely when changing rimage target,

  Current value="", Ignored value="cnl"
Call Stack (most recent call first):
  /projects/github/ncs/zephyr/cmake/extensions.cmake:829 (zephyr_check_cache)
  /projects/github/ncs/zephyr/boards/xtensa/intel_adsp_cavs18/board.cmake:6 (board_set_rimage_target)
  /projects/github/ncs/zephyr/cmake/app/boilerplate.cmake:627 (include)
  /projects/github/ncs/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:24 (include)
  /projects/github/ncs/zephyr/share/zephyr-package/cmake/ZephyrConfig.cmake:40 (include_boilerplate)
  CMakeLists.txt:5 (find_package)
```

@MaureenHelm so this should be fixed.
It has nothing to do with passing extra arguments later.

**Comment by marc-hb** [comment]
@cfriedt can you share a few words about why #40642 was closed and moved here?

**Comment by cfriedt** [comment]
@marc-hb - it seemed like more of a discussion than an enhancement request or a bug report. If that's not the case, then please follow up with @MaureenHelm or create an issue using one of the existing templates.

**Comment by marc-hb** [comment]
It was a real bug with very well defined reproduction steps. The title was misleading sorry about that. I didn't use a bug template because there was a ton of information behind the links and a very well identified commit that introduced the regression but you're right: even that shouldn't grant any process exception otherwise anything goes; apologies.

**Comment by tejlmand** [comment, marked as answer]
created new bug here: https://github.com/zephyrproject-rtos/zephyr/issues/40808
